### PR TITLE
Use BrickMoney for Smarty modifier 

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Switchings crmMoney to use BrickMoney

Before
----------------------------------------
Custom handling

After
----------------------------------------
Relevant package

Technical Details
----------------------------------------
Note that this will need testing in various currency & locale configs

Note that using ```CRM_Core_I18n::singleton()->getLocale()``` for the locale might have some
edge cases.

If I'm in NZ paying in Euros should I see the decimal in the NZ place or the Euro
place. Backoffice users should not be confused by commas etc.  I haven't tested these
variations yet so flagging them as things to consider during review stage

Comments
----------------------------------------
Needs a good tyre-kicking

### CODE REVIEW

File: CRM/Core/Smarty/plugins/modifier.crmMoney.php [hunk 1]
```diff
@@ -31,5 +27,11 @@
  * @throws \CRM_Core_Exception
  */
 function smarty_modifier_crmMoney($amount, $currency = NULL) {
-  return CRM_Utils_Money::format($amount, $currency);
+  if (!$amount) {
+    return $amount;
+  }
+  $currency = $currency ?? CRM_Core_Config::singleton()->defaultCurrency;
+  $money = Money::of($amount, $currency, new DefaultContext(), RoundingMode::CEILING);
+  $formatter = new \NumberFormatter(CRM_Core_I18n::singleton()->getLocale(), \NumberFormatter::CURRENCY);
```
Comment (seamuslee001) [inline CRM/Core/Smarty/plugins/modifier.crmMoney.php:35]: @eileenmcnaughton is there a reason why you didn't do something like "echo $money->formatTo('en_US');" instead?

Comment (eileenmcnaughton) [reply CRM/Core/Smarty/plugins/modifier.crmMoney.php:35]: I read a different part of the docs? We have $currency passed in - I 'guessing' locale

**Comment by mattwire** [comment]
> If I'm in NZ paying in Euros should I see the decimal in the NZ place or the Euro
place.

@eileenmcnaughton The whole system needs to display currency using one or the other format for decimal separator, not a mix of both.

**Comment by seamuslee001** [comment]
@mattwire @eileenmcnaughton is the decimalSeparator / thousandsSeparator more of a Locale thing rather than a currency thing? If that is the case as long as we are always outputting based on the System Locale then it should be consistent right?

**Comment by eileenmcnaughton** [comment]
@mattwire so if I have a site in NZ but we have a page with Euro currency if should display 

EUR 22.50

NOT 

EUR 22,55

- that's what we currently & this doesn't change it -   but I'm pretty sure WMF donation forms present the currency formatting appropriate to the currency

**Comment by eileenmcnaughton** [comment]
Test fails are because we are getting

> $100.00

instead of
> $ 100.00

ie the extra space is removed

**Comment by eileenmcnaughton** [comment]
Note amazon doesn't have the extra space...

Screen Shot 2020-06-19 at 8 26 16 PM

**Comment by mattwire** [comment]
@eileenmcnaughton @seamuslee001 I agree. The currency formatting is a locale thing not a currency thing. Eg. a currency specifies a number of decimal places and a symbol. But the locale specifies how to display currencies:
- Whether to have a space between the symbol and value.
- Whether to display symbol after or before the value.
- Whether to use , or . as decimal separator.

So someone viewing a page in French or Spanish would expect to see 24,99€, in German you would see 24,99 € and in English you'd see €24.99.

(I know we don't currently implement some of those formatting features above but we might in future).

**Comment by eileenmcnaughton** [comment]
@mattwire yeah - it's a bit tangental to this - since it makes no change but something to think about since it seems we could improve on it in the future.

It's not an 'either/or' though - brickmoney -  uses both pieces of information. For example the currency affects the symbol and the number of decimal places, the locale affects the order etc.

I'll fix those tests but will need some varied user testing on this.

@seamuslee001 I'm happy to change the syntax if you think the other is bettter

**Comment by seamuslee001** [comment]
@eileenmcnaughton I would given i think that is the standard brickmoney format to print out different locale if you can

**Comment by eileenmcnaughton** [comment]
@seamuslee001 so you are implying you think it actually does something different rather than is a different option?

**Comment by seamuslee001** [comment]
@eileenmcnaughton I'm saying that using ->toLocale() is the standard BrickMoney way of outputting the relevant currency in a different locale rather than needing NumberFormatter directly

**Comment by eileenmcnaughton** [comment]
Closing this as it's more of a prototype / discussion piece than the final

**Pull request closed without merging**